README.md
fbp-spec
========

A runtime-independent test framework for Flow Based Programming (FBP) component and graphs,
using declarative, data-driven testing.

One can use fbp-spec to do testing at multiple levels,
each approximately corresponding to the different architectural levels of Flow Based Programming:

* Unit (FBP component/subgraph)
* Integration (FBP graph)
* System (FBP runtime)

## Status:

**In production**

* Used by several projects, including [Ingress Table](https://github.com/c-base/ingress-table) and [BIG IoT NoFlo bridge](https://github.com/flowhub/bigiot-bridge)
* Tested with several FBP runtimes: [NoFlo](https://noflojs.org), [MicroFlo](https://microflo.org), [Python example](https://github.com/flowbased/protocol-examples), [MsgFlo](http://msgflo.org)
* Runners available for contious integration (CLI, Mocha) and interactively (in [Flowhub](https://flowhub.org))

## Purpose & Scope

Note: `fbp-spec` is intended for use by application and component-library developers.

The following is considered out-of-scope:

* Testing conformance with the FBP protocol. Instead use [fbp-protocol](https://github.com/flowbased/fbp-protocol)
* Testing an FBP runtime/engine itself. Instead use a testing framework for your particular runtime language/environment.

## License

The [MIT license](./LICENSE.md)

# Usage

## Installing

Set up fbp-spec as an NPM dependency

    npm install --save-dev fbp-spec

or, install it globally. Useful if you just want the commandline tool.

    npm install -g fbp-spec

## Writing tests

Each declared test suite loads an FBP component (or graph) fixture,
and runs a set of test cases by sending a set of input data
to input ports and verifying the output data against the expected results.

```YAML
name: "Simple example of passing tests"
topic: "core/Repeat"
fixture:
 type: 'fbp'
 data: |
  INPORT=it.IN:IN
  OUTPORT=f.OUT:OUT
  it(core/Repeat) OUT -> IN f(core/Repeat)

cases:
-
  name: 'sending a boolean'
  assertion: 'should repeat the same'
  inputs:
    in: true
  expect:
    out:
      equals: true
-
  name: 'sending a number'
  assertion: 'should repeat the same'
  inputs:
    in: 1000
  expect:
    out:
      equals: 1000
```

#### Multiple ports

You can send data to multiple inports and check expectations on multiple ports per testcase:

```YAML
-
  name: '1 active track toggled high'
  assertion: 'should give value1 color'
  inputs:
    tracks: 1
    animation: [
      0, # track idx
      "0xEE00EE", # val0
      "0xAA00AA", # val1
      200, # period
      50, # dutycycle
      0, # offset
      500 ] # duration
    clock: 250
  expect:
    clock:
     equals: 250
    value:
     equals: [0, 0x00AA] # FIXME: truncated
```

#### Sequence of packets

For testing components with state, you can sending multiple input packets
[in sequence](https://github.com/flowbased/fbp-spec/blob/master/examples/sequences.yaml).

```YAML
-
  name: 'sequence of data using spacy notation'
  assertion: 'should pass'
  inputs:
    -
      in: true
    -
      in: false
  expect:
    - 
      out:
        equals: true
    -
      out:
        equals: false 
```

#### Extract data using path

With `path` you can specify a [JSONPath](http://goessner.net/articles/JsonPath/)
to extract the piece(s) of data the assertions will be ran against:

```YAML
-
  name: 'select single value'
  assertion: 'should pass'
  inputs:
    in: { outer: { inner: { foo: 'bar' } } }
  expect:
    out:
      path: '$.outer.inner.foo'
      equals: 'bar'
-
  name: 'selecting many correct values'
  assertion: 'should pass'
  inputs:
    in:
      outer:
        first: { foo: 'bar' }
        second: { foo: 'bar' }
  expect:
    out:
      path: '$.outer.*.foo'
      equals: 'bar'
```

#### Skipping tests

Setting `skip` property on a testcase or suite, will cause it to not be ran.
Should contain a message of the reason for skipping.

```YAML
-
  name: 'a test that is skipped'
  assertion: 'will not be ran'
  inputs:
    in: 1000
  expect:
    out:
      equals: 1000
  skip: 'not implemented yet'
```

#### Using fixtures

One can use testing-specific components in the fixture, to simplify
driving the unit under test with complex inputs and performing complex assertions.

```YAML
fixture:
 type: 'fbp'
 data: |
  INPORT=imagename.IN:NAME
  INPORT=testee.PARAM:PARAM
  INPORT=reference.IN:REFERENCE
  OUTPORT=compare.OUT:SIMILARITY

  generate(test/GenerateTestImage) OUT -> IN testee(my/Component)
  testee OUT -> ACTUAL compare(test/CompareImage)
  reference(test/ReadReferenceImage) OUT -> REFERENCE compare
cases:
-
  name: 'testing complex data with custom components fixture'
  assertion: 'should pass'
  inputs:
    name: someimage
    param: 100
    reference: someimage-100-result
  expect:
    similarity:
      above: 0.99
```

#### Supported assertions

Instead of `equals` you can use any of the supported assertion predicates. Examples include:

    type
    above
    below
    contains
    haveKeys
    includeKeys

For a full set of assertions, see [the schema](https://github.com/flowbased/fbp-spec/blob/master/schemata/expectation.yaml)

#### More

A comprehensive set of examples can be found under [./examples](./examples).
For the detailed definition of the dataformat for tests, see [schemata/](./schemata/).


## Running tests with fbp-spec commandline tool

The simplest and most universal way of running tests is with the `fbp-spec` commandline tool.

    $ fbp-spec --address ws://localhost:3333 examples/multisuite-failandpass.yaml

    MultiSuite, failing tests
      sending a boolean with wrong expect
        should fail: ✗ Error: expected true to deeply equal false
      sending a number with wrong expect
        should fail: ✗ Error: expected 1000 to deeply equal 1003
    MultiSuite, passing tests
      sending a boolean
        should repeat the same: ✓
      sending a number
        should repeat the same: ✓

The `--command` options can be used to specify a command which will start the runtime under test:

    fbp-spec --command "python2 protocol-examples/python/runtime.py"

It sets the exit status to non-zero on failure, so is suitable for integrating into a `Makefile` or similar.

## Running tests by integrating with Mocha

[Mocha](http://mochajs.org/) iss a popular test runner framework for JavaScript/CoffeeScript on browser and node.js.

Since fbp-spec communicates with your runtime over a network protocol,
you can use this also when your project is not JavaScript-based.
The Mocha runner is for instance [used in microflo-core](https://github.com/microflo/microflo-core/blob/master/spec/ComponentTests.coffee)
to test C++ components for microcontrollers & embedded devices.

You can have your fbp-spec tests run in Mocha by calling the `fbpspec.mocha.run()` function, in a file which is
executed with the standard Mocha runner. Eg. `mocha --reporter spec tests/fbpspecs.js`

    // fbpspecs.js
    fbpspec = require('fbp-spec');

    rt = {
      protocol: "websocket",
      address: "ws://localhost:3569",
      secret: "py3k", // Optional. If needed to connect/authenticate to runtime
      command: 'python2 protocol-examples/python/runtime.py' // Optional. Can be used to start runtime automatically
    };
    fbpspec.mocha.run(rt, './examples/simple-passing.yaml', { starttimeout: 1000 });

The tests can be specified as a list of files, or directories.
You can use the standard `grep` option of Mocha to run only some tests.

For CoffeScript example, see [./spec/mocha.js](./spec/mocha.js).

## Running tests interactively in Flowhub

[Flowhub](http://app.flowhub.io) IDE (version 0.11 and later) has integrated support for fbp-spec. No installation is required.

* Open existing project, or create a new one
* Open a component, and write/copypaste in a test in the `Tests` panel
* Ensure you have a runtime set up, and connected

When you make changes to your project (components,graphs) or tests, Flowhub will now automatically (re-)run your tests.
You can see the status in the top-right corner. Clicking on it brings up more details.

## Generating tests programatically

The test-format defined by fbp-spec is fairly generic and versatile. It is intended primarily as
a format one directly specifies tests in, but can also be generated from other sources.

Sometimes data-driven testing, one does a large amount of very similar tests,
with multiple test-cases per set of input data.
By capturing only the unique parts of testcases in a specialied data-structure (JSON, YAML, etc),
and then transforming this into standard `fbp-spec` files with some code, adding/removing
cases becomes even easier.
For instance in `imgflo-server`, [testcases](https://github.com/jonnor/imgflo-server/blob/master/spec/graphtests.yaml)
can be defined by providing a name, an URL and a reference result (a file with naming convention based on name).

Similarly, one can generate testcases using fuzzing, schema-based, model-based or similar tools.

## Integrating test runner in an application

The test runner code is accessible as a JavaScript library,
and can be integrated into other apps (like Flowhub does).
See examples of [commandline](./lib/cli.js) and webappp usage.

## Add supporting for a new runtime

You need to implement the [FBP network protocol](https://github.com/flowbased/fbp-protocol).
At least the `protocol:runtime`, `protocol:graph`, and `protocol:network` capabilities are required.

All transports supported by [fbp-protocol-client]((https://github.com/flowbased/fbp-protocol))
are supported by fbp-spec, including WebSocket, WebRTC, and iframe/postMessage.

fbp-spec is intended to be used with flow-based and dataflow-programming,
but might be useful also outside these programming paradigms. Try it out!

## Writing a test runner in another language

As long as you stay compatible with the [fbp-spec testformat](./schemata/)
and [FBP protocol](http://noflojs.org/documentation/protocol/),
you can implement a compatible runner in any programming language.

You can consider the fbp-spec code (in CoffeeScript) as a reference implementation.



CHANGES.md
# 0.8.0 - released 11.11.2020

* Added a `--trace` command-line flag to capture Flowtraces for fbp-spec runs. The Flowtraces will be stored into a `.flowtrace` folder under the project

# 0.7.1 - released 22.10.2020

* Fixed issue with loading spec files

# 0.7.0 - released 22.10.2020

* Added a `--component-tests` command-line flag to control whether fbp-spec loads additional component tests from the runtime
* Ported from CoffeeScript to modern JavaScript

# 0.6.8 - released 09.09.2020

* Made fbp-spec ignore failing getSource requests
* If network setup fails, teardown for the network will also be skipped

# 0.6.7 - released 01.09.2020

* Updated dependencies to include latest FBP Protocol Client and Schemas
* Updated browser build to use a modern Babel version

# 0.6.2 - released 31.03.2018

* Fixed a race condition with synchronous networks where we expect multiple output packets

# 0.6.1 - released 31.03.2018

* Fixed issue where received bracket IPs were causing fbp-spec to not see the actual data packet

# 0.6.0 - released 30.03.2018

* Switched FBP Protocol communications with the runtime to use the new [fbp-client](https://github.com/flowbased/fbp-client) library
* All messages to and from runtime are now validated against the FBP Protocol schema
* Permissions are now validated against capabilities advertised by the runtime
* Tests now fail if runtime breaks the connection (for example, if runtime crashes mid-run)
* Tests now fail if runtime sends a `network:error` message
* Tests now fail if runtime sends a `network:processerror` message
* Tests now fail if runtime sends a message to an exported `error` outport if message to that port was expected
* All FBP Protocol operations now time out if there is no response

# 0.5.0 - released 17.11.2017

* Updated to fbp parser 1.7 which introduces consistency validations for graphs. This can cause some tests to break due to them having incorrectly defined nodes in them

Bugfixes:

* We actually catch and show FBP parsing errors now

# 0.4.1 - released 07.11.2017

Breaking changes

* Require ES6 support. Node.js 6+ or a modern browser.
If support for older environments are needed, use Babel to transpile.

Internal changes

* Now using CoffeeScript 2

# 0.3.0 - released 06.11.2017

* Updated to fbp-protocol-client 0.2.x, without support for MicroFlo transport

# 0.2.3 - released 04.10.2017

Bugfixes

* Fixed triggering Mocha error in start/stop due to returning Promise. Regression since 0.2.2.

# 0.2.2 - released 03.10.2017

* Automatically retry connection while during setup until `startTimeout`

# 0.2.0 - released 19.02.2017

* Updated to Mocha 3.x.
Can break tests in some cases where Promise is returned and callback is used at same time.

Internal changes

* Updated all dependencies to latest version

# 0.1.16 - released 11.10.2016

Features

* Several conveniences added in .fbp DSL from `fbp 1.5`

# 0.1.8 - released 20.04.2016

* `testsuite.validate()`: Fix not catching invalid files, due to wrong schema lookup
* Add support for sequences of inputs/expect. Allows testing stateful components/graphs/systems.

# 0.1.7 - released 17.04.2016

* Make `fixture` optional. If not specified, will be auto-created from component/graph in `topic`

# 0.1.6 - released 14.04.2016

* Some improvements to experimental `fbp-spec-mocha`

# 0.1.4 - released 07.05.2016

* Send `network:output` FBP messages to `console.log()`, for easier debugging of problems on runtime side.
* Added experimental `fbp-spec-mocha` tool, for running Mocha tests over fbp-spec protocol.

# 0.1.3 - released 27.08.2015

First useful release

* Initial YAML testformat fully defined
* Tested with NoFlo, MicroFlo and Python example runtimes
* `fbp-spec` commandline tool for running
* `fbpspec.mocha.run` for integration in a Mocha testsuite
* Simple set of UI widgets, with experimental integration in noflo-ui

# 0.0.1 - released 17.04.2015

Proof of concept

* Could run a MicroFlo testcase by talking over FBP runtime protocol


lib/runner.js
const fbp = require('fbp');
const fbpClient = require('fbp-client');
const debug = require('debug')('fbp-spec:runner');
const debugReceived = require('debug')('fbp-spec:runner:received');
const Promise = require('bluebird');
const { FlowtraceRecorder } = require('flowtrace');
const common = require('./common');
const protocol = require('./protocol');
const testsuite = require('./testsuite');
const expectation = require('./expectation');

function debugReceivedMessages(client) {
  client.on('signal', ({ proto, command, payload }) => debugReceived(proto, command, payload));
}

function synthesizeTopicFixture(topic, components, callback) {
  debug('synthesizing fixture', topic);
  // export each of the ports for topic component
  // return a FBP graph?
  const component = components[topic];
  if (!component) {
    callback(new Error(`Could not find component for topic: ${topic}`));
    return;
  }

  const graph = {
    properties: {},
    inports: {},
    outports: {},
    processes: {},
    connections: [],
  };

  const processName = 'testee';
  graph.processes[processName] = { component: topic };

  component.inPorts.forEach((port) => {
    const portName = port.id;
    graph.inports[portName] = {
      process: processName,
      port: portName,
    };
  });
  component.outPorts.forEach((port) => {
    const portName = port.id;
    graph.outports[portName] = {
      process: processName,
      port: portName,
    };
  });

  // Sanity checking if this is usable as a fixture
  if (Object.keys(graph.outports).length < 1) {
    callback(new Error(`Component '${topic}' used as fixture has no outports`));
    return;
  }
  if (Object.keys(graph.inports).length < 1) {
    callback(new Error(`Component '${topic}' used as fixture has no inports`));
    return;
  }

  callback(null, graph);
}

// @context should have .components = {} property
function getFixtureGraph(context, suite, callback) {
  // TODO: follow runtime for component changes

  const hasComponents = (s) => (s.components != null) && (typeof s.components === 'object') && Object.keys(s.components).length;

  function updateComponents(cb) {
    if (hasComponents(context)) {
      cb(null);
      return;
    }
    protocol.getComponents(context.client, (err, components) => {
      if (err) {
        cb(err);
        return;
      }
      context.components = components;
      cb(null);
    });
  }

  updateComponents((err) => {
    if (err) {
      callback(err);
      return;
    }

    if (!suite.fixture) {
      synthesizeTopicFixture(suite.topic, context.components, callback);
      return;
    }

    if (suite.fixture.type === 'json') {
      let graph;
      try {
        graph = JSON.parse(suite.fixture.data);
      } catch (error) {
        callback(error);
        return;
      }
      callback(null, graph);
      return;
    }
    if (suite.fixture.type === 'fbp') {
      let graph;
      try {
        graph = fbp.parse(suite.fixture.data);
      } catch (error) {
        callback(error);
        return;
      }

      if (!graph.properties) {
        graph.properties = {};
      }
      callback(null, graph);
      return;
    }
    callback(new Error(`Unknown fixture type ${suite.fixture.type}`));
  });
}

function sendMessageAndWait(client, currentGraph, inputData, expectData, callback) {
  const observer = client.observe((s) => (s.protocol === 'runtime') && (s.command === 'packet') && (s.payload.graph === currentGraph));

  function signalsToReceived(signals) {
    const received = {};
    signals.forEach((signal) => {
      received[signal.payload.port] = signal.payload.payload;
    });
    return received;
  }

  function checkSuccess(s) {
    // We're only interested in response packets
    if ((s.protocol !== 'runtime') || (s.command !== 'packet')) { return false; }
    // Check that is for the current graph under test
    if (s.payload.graph !== currentGraph) { return false; }
    // We're only interested in data IPs, not brackets
    if (s.payload.event !== 'data') { return false; }
    // Get signals received until this message
    const receivedSignals = observer.signals.slice(0, observer.signals.indexOf(s) + 1);
    const received = signalsToReceived(receivedSignals);
    const result = (Object.keys(received).length === Object.keys(expectData).length);
    return result;
  }
  function checkFailure(s) {
    if ((s.protocol === 'network') && (s.command === 'error')) {
      // network:error should imply failed test
      if (s.payload.graph && (s.payload.graph !== currentGraph)) { return false; }
      return true;
    }
    if ((s.protocol === 'network') && (s.command === 'processerror')) {
      // network:processerror should imply failed test
      if (s.payload.graph && (s.payload.graph !== currentGraph)) { return false; }
      return true;
    }
    if ((s.protocol === 'runtime') && (s.command === 'packet')) {
      // Output packet, see if it is an unexpected error
      // Check that is for the current graph under test
      if (s.payload.graph !== currentGraph) { return false; }
      // We're only interested in data IPs, not brackets
      if (s.payload.event !== 'data') { return false; }
      // We only care for packets to error port
      if (s.payload.port !== 'error') { return false; }
      // We only care if spec isn't expecting errors
      if (typeof expectData.error !== 'undefined') { return false; }
      return true;
    }
    return false;
  }

  // send input packets
  const sendPackets = Promise.promisify(protocol.sendPackets);
  Promise.resolve()
    .then(() => sendPackets(client, currentGraph, inputData))
    .then(() => observer.until(checkSuccess, checkFailure))
    .then((signals) => signalsToReceived(signals))
    .nodeify(callback);
}

function needsSetup(suite) {
  const notSkipped = suite.cases.filter((c) => !c.skip);
  return notSkipped.length > 0;
}

class Runner {
  constructor(client, options = {}) {
    this.client = client;
    this.currentGraphId = null;
    this.components = {};
    this.parentElement = null;
    this.options = options;
    this.recorder = null;
    if ((this.options.connectTimeout == null)) { this.options.connectTimeout = 5 * 1000; }
    if ((this.options.commandTimeout == null)) { this.options.commandTimeout = 3 * 1000; }
  }

  prepareClient(callback) {
    if ((this.client.protocol != null) && (this.client.address != null)) {
      // is a runtime definition
      Promise.resolve()
        .then(() => fbpClient(this.client, { commandTimeout: this.options.commandTimeout }))
        .then((client) => {
          this.client = client;

          if (this.parentElement && (client.definition.protocol === 'iframe')) {
            // We need to set up the parent element in this case
            client.transport.setParentElement(this.parentElement);
          }

          return client;
        })
        .nodeify(callback);
      return;
    }
    // This is a client instance
    callback(null, this.client);
  }

  // TODO: check the runtime capabilities before continuing
  connect(callback) {
    debug('connect');

    this.prepareClient((err) => {
      if (err) {
        callback(err);
        return;
      }

      debugReceivedMessages(this.client);
      this.client.on('network', ({ command, payload }) => {
        if ((command === 'output') && payload.message) {
          console.log(payload.message);
        }
      });

      const timeBetweenAttempts = 500;
      const attempts = Math.floor(this.options.connectTimeout / timeBetweenAttempts);
      const isOnline = () => {
        const connected = this.client.isConnected();
        if (connected) {
          return Promise.resolve();
        }
        return Promise.reject(new Error('Not connected to runtime'));
      };
      const tryConnect = () => {
        debug('trying to connect');
        return this.client.connect();
      };
      common.retryUntil(tryConnect, isOnline, timeBetweenAttempts, attempts)
        .then(() => this.checkCapabilities(['protocol:graph', 'protocol:network', 'protocol:runtime']))
        .nodeify(callback);
    });
  }

  disconnect(callback) {
    debug('disconnect');

    if (!(this.client != null ? this.client.isConnected() : undefined)) {
      callback();
      return;
    }

    Promise.resolve()
      .then(() => this.client.disconnect())
      .nodeify(callback);
  }

  checkCapabilities(capabilities) {
    if (!this.client.isConnected()) {
      return Promise.reject(new Error('Not connected to runtime'));
    }
    if (!this.client.definition
      || !this.client.definition.capabilities
      || !this.client.definition.capabilities.length) {
      return Promise.reject(new Error('Runtime provides no capabilities'));
    }
    for (let i = 0; i < capabilities.length; i += 1) {
      const capability = capabilities[i];
      if (this.client.definition.capabilities.indexOf(capability) === -1) {
        return Promise.reject(new Error(`Runtime doesn't provide ${capability}`));
      }
    }
    return Promise.resolve();
  }

  setupSuite(suite, callback) {
    debug(`setup suite "${suite.name}"`);
    if (this.options.traceCallback && !this.recorder) {
      this.recorder = new FlowtraceRecorder(this.client);
    }
    if (!needsSetup(suite)) {
      callback(null);
      return;
    }

    if (!this.client.isConnected()) {
      callback(new Error('Disconnected from runtime'));
      return;
    }

    getFixtureGraph(this, suite, (err, graph) => {
      if (err) {
        callback(err);
        return;
      }
      protocol.sendGraph(this.client, graph, (sendErr, graphId) => {
        if (sendErr) {
          callback(sendErr);
          return;
        }
        this.currentGraphId = graphId;
        if (this.recorder) {
          this.recorder.start(graphId)
            .then(() => {
              protocol.startNetwork(this.client, graphId, callback);
            }, callback);
          return;
        }
        protocol.startNetwork(this.client, graphId, callback);
      });
    });
  }

  teardownSuite(suite, callback) {
    debug(`teardown suite "${suite.name}"`);
    if (this.recorder && this.currentGraphId) {
      // FIXME: This returns a Promise that we should wait for
      this.recorder.stop(this.currentGraphId);
    }
    if (!needsSetup(suite)) {
      callback(null);
      return;
    }

    if (!this.client.isConnected()) {
      callback(new Error('Disconnected from runtime'));
      return;
    }

    if (!this.currentGraphId) {
      // Graph was never successfully set up, so no need to stop
      callback();
      return;
    }

    // FIXME: also remove the graph. Ideally using a 'destroy' message in FBP protocol
    protocol.stopNetwork(this.client, this.currentGraphId, callback);
  }

  runTest(testcase, callback) {
    debug(`runtest "${testcase.name}"`);

    if (!this.client.isConnected()) {
      callback(new Error('Disconnected from runtime'));
      return;
    }

    // XXX: normalize and validate in testsuite.js instead?
    const inputs = common.isArray(testcase.inputs) ? testcase.inputs : [testcase.inputs];
    const expects = common.isArray(testcase.expect) ? testcase.expect : [testcase.expect];
    const sequence = [];
    for (let i = 0, end = inputs.length, asc = end >= 0;
      asc ? i < end : i > end;
      asc ? i += 1 : i -= 1) {
      sequence.push({
        inputs: inputs[i],
        expect: expects[i],
      });
    }

    const sendWait = (data, cb) => sendMessageAndWait(
      this.client,
      this.currentGraphId,
      data.inputs,
      data.expect,
      cb,
    );
    common.asyncSeries(sequence, sendWait, (err, actuals) => {
      if (this.options.traceCallback && this.currentGraphId && this.recorder) {
        this.options.traceCallback(
          this.currentGraphId,
          testcase.name,
          this.recorder.dump(this.currentGraphId),
        );
      }
      if (err) {
        callback(err);
        return;
      }
      actuals.forEach((r, idx) => {
        sequence[idx].actual = r;
      });
      callback(null, sequence);
    });
  }
}

// TODO: should this go into expectation?
function checkResults(results) {
  let result;
  const actuals = results.filter((r) => r.actual != null);
  const expects = results.filter((r) => r.expect != null);
  if (actuals.length < expects.length) {
    result = {
      passed: false,
      error: new Error(`Only got ${actuals.length} output messages out of ${expects.length}`),
    };
    return result;
  }

  const mappedResults = results.map((res) => {
    res.error = null;
    try {
      expectation.expect(res.expect, res.actual);
    } catch (e) {
      // FIXME: only catch actual AssertionErrors
      res.error = e;
    }
    return res;
  });

  const failures = mappedResults.filter((r) => r.error);
  if (failures.length === 0) {
    result = { passed: true };
  } else if (expects.length === 1) {
    result = { error: failures[0].error };
  } else if ((expects.length > 1) && (failures.length === 1)) {
    const index = mappedResults.findIndex((r) => r.error);
    const failed = mappedResults[index];
    result = { error: new Error(`Expectation ${index} of sequence failed: ${failed.error.message}`) };
  } else {
    const errors = mappedResults.map((r) => (r.error != null ? r.error.message : undefined) || '');
    result = { error: new Error(`Multiple failures in sequence: ${errors}`) };
  }

  return result;
}

function runTestAndCheck(runner, testcase, callback) {
  if (testcase.skip) {
    // TODO: pass some skipped state? its indirectly in .skip though
    callback(null, { passed: true });
    return;
  }

  // XXX: normalize and validate in testsuite.js instead?
  const inputs = common.isArray(testcase.inputs) ? testcase.inputs : [testcase.inputs];
  const expects = common.isArray(testcase.expect) ? testcase.expect : [testcase.expect];
  if (inputs.length !== expects.length) {
    callback(null, {
      passed: false,
      error: new Error(`Test sequence length mismatch. Got ${inputs.length} inputs and ${expects.length} expectations`),
    });
    return;
  }

  runner.runTest(testcase, (err, results) => {
    let result;
    if (err) {
      // Map error to a test failure
      result = {
        passed: false,
        error: err,
      };
      callback(null, result);
      return;
    }
    result = checkResults(results);
    if (result.error) {
      result.passed = false;
    }
    callback(null, result);
  });
}

function runSuite(runner, suite, runTest, callback) {
  if (suite.skip) {
    // TODO: pass some skipped state? its indirectly in .skip though
    callback(null, suite);
    return;
  }

  runner.setupSuite(suite, (err) => {
    debug('setup suite', err);
    if (err) {
      callback(err, suite);
      return;
    }

    common.asyncSeries(suite.cases, runTest, (runErr) => {
      debug('testrun complete', runErr);
      if (runErr) {
        runner.teardownSuite(suite, (teardownErr) => {
          debug('teardown suite', teardownErr);
          callback(runErr, suite);
        });
        return;
      }

      runner.teardownSuite(suite, (teardownErr) => {
        debug('teardown suite', teardownErr);
        callback(teardownErr, suite);
      });
    });
  });
}

function loadComponentSuites(componentTests, callback) {
  function loadYAML(source, cb) {
    testsuite.loadYAML(source, (err, suite) => {
      if (err) {
        // ignore, could be non fbp-spec test
        // TODO: include tests type in FBP protocol, so we know whether this is error or legit
        cb(null, null);
        return;
      }
      cb(null, suite);
    });
  }
  const tests = Object.keys(componentTests).map((name) => componentTests[name]);
  common.asyncSeries(tests, loadYAML, (err, suites) => {
    if (err) {
      callback(err);
      return;
    }
    callback(null, suites.filter((s) => s !== null).reduce((a, b) => a.concat(b), []));
  });
}

exports.getComponentSuites = function getComponentSuites(runner, callback) {
  protocol.getCapabilities(runner.client, (err, caps) => {
    if (err) {
      callback(err);
      return;
    }
    if (caps.indexOf('component:getsource') === -1) {
      callback(null, []);
      return;
    }

    protocol.getComponentTests(runner.client, (testsErr, tests) => {
      if (testsErr) {
        callback(testsErr);
        return;
      }
      loadComponentSuites(tests, (loadErr, suites) => {
        if (loadErr) {
          callback(loadErr);
          return;
        }
        debug('get component suites', tests.length, suites.length);
        callback(null, suites);
      });
    });
  });
};

// will update each of the testcases in @suites
// with .passed and .error states as tests are ran
function runAll(runner, suites, updateCallback, doneCallback) {
  function runTest(testcase, callback) {
    function done(error) {
      updateCallback(suites);
      callback(error);
    }

    runTestAndCheck(runner, testcase, (err, results) => {
      // ignore error to not bail out early
      const tcase = testcase;
      Object.keys(results).forEach((key) => {
        const val = results[key];
        tcase[key] = val;
      });
      if (testcase.error) {
        tcase.error = testcase.error.message;
      }
      debug(`ran test "${testcase.name}"`, testcase.passed, err);
      done(null);
    });
  }

  const runOneSuite = (suite, cb) => runSuite(runner, suite, runTest, cb);

  debug('running suites', suites.map((s) => s.name));
  common.asyncSeries(suites, runOneSuite, (err) => doneCallback(err));
}

exports.Runner = Runner;
exports.runAll = runAll;
exports.runTestAndCheck = runTestAndCheck;
